File: src/main/java/com/dec/decisland/datagen/ModLangProvider.java
package com.dec.decisland.datagen;

import com.dec.decisland.DecIsland;
import com.dec.decisland.block.ModBlocks;
import com.dec.decisland.item.ModCreativeModeTabs;
import com.dec.decisland.item.ModItems;
import net.minecraft.data.PackOutput;
import net.minecraft.world.item.Item;
import net.minecraft.core.Holder;
import net.minecraft.world.level.block.Block;
import net.neoforged.neoforge.common.data.LanguageProvider;

// 通用语言提供者基类
public abstract class ModLangProvider extends LanguageProvider {
    protected final String locale;
    protected ModLangProvider(PackOutput output, String locale) {
        super(output, DecIsland.MOD_ID, locale);
        this.locale = locale;
    }

    @Override
    protected void addTranslations() {
        // 物品翻译
        ModItems.getItemConfigs().forEach(config -> {
            Item item = ModItems.getItemByConfig(config);

            if (item != null && config.langMap != null) {
                String translation = config.langMap.get(this.locale);
                if (translation != null) {
                    add(item, translation);
                }
            }
        });

        // 方块翻译
        ModBlocks.getBlockConfigs().forEach(config -> {
            Block block = ModBlocks.getBlockByConfig(config).value();

            if (block != null && config.langMap != null) {
                String translation = config.langMap.get(this.locale);
                if (translation != null) {
                    add(block, translation);
                }
            }
        });

        // 物品组翻译
        ModCreativeModeTabs.getTabConfigs().forEach(config -> {
            if (config.langMap == null) return;
            String translation = config.langMap.get(this.locale);
            if (translation == null) return;
            add("itemGroup." + config.name, translation);
        });
    }
}

File: src/main/java/com/dec/decisland/datagen/ModLangProvider.kt
package com.dec.decisland.datagen

import com.dec.decisland.DecIsland
import com.dec.decisland.block.ModBlocks
import com.dec.decisland.effect.ModEffects
import com.dec.decisland.item.ModCreativeModeTabs
import com.dec.decisland.item.ModItems
import com.dec.decisland.datagen.Lang.BedrockLangResources
import net.minecraft.data.PackOutput
import net.neoforged.neoforge.common.data.LanguageProvider

abstract class ModLangProvider protected constructor(
    output: PackOutput,
    protected val locale: String,
) : LanguageProvider(output, DecIsland.MOD_ID, locale) {
    override fun addTranslations() {
        val translations = linkedMapOf<String, String>()

        // Layer 2: Bedrock entries are imported first and act as the global fallback.
        translations.putAll(BedrockLangResources.javaTranslations(locale))

        // Layer 1: explicit Java-side definitions override Bedrock when both are present.
        ModItems.getItemConfigs().forEach { config ->
            val translation = config.langMap[locale] ?: BedrockLangResources.itemName(locale, config.name)
            if (translation != null) {
                translations["item.${DecIsland.MOD_ID}.${config.name}"] = translation
            }
        }

        ModBlocks.getBlockConfigs().forEach { config ->
            val translation = config.langMap[locale] ?: BedrockLangResources.blockName(locale, config.name)
            if (translation != null) {
                translations["block.${DecIsland.MOD_ID}.${config.name}"] = translation
            }
        }

        ModCreativeModeTabs.getTabConfigs().forEach { config ->
            val translation = config.langMap[locale] ?: return@forEach
            translations["itemGroup.${config.name}"] = translation
        }

        when (locale) {
            "en_us" -> translations[ModEffects.DIZZINESS.get().descriptionId] = "Dizziness"
            "zh_cn" -> translations[ModEffects.DIZZINESS.get().descriptionId] = "眩晕"
        }

        translations.forEach(::add)
    }
}
